### PR TITLE
USB HID minimal implementation

### PR DESCRIPTION
**Pull request by Ollrogge**
### Contribution description
Minimal implementation of USB HID using the USBUS. 
This implementation is needed as base in order to implement the Client to Authenticator Protocol (CTAP), which is part of FIDO2, in RIOT.

### Testing procedure
* tests/usbus_hid

### CODE REVIEW

File: tests/usbus_hid/main.c
Comment (PeterKietzmann) [inline tests/usbus_hid/main.c:42]: Is this function actually documented somewhere?

File: tests/usbus_hid/main.c
Comment (PeterKietzmann) [inline tests/usbus_hid/main.c:9]: FIDO**2** right?

File: tests/usbus_hid/main.c
Comment (PeterKietzmann) [inline tests/usbus_hid/main.c:1]: Copyright header missing.

File: tests/usbus_hid/main.c
Comment (PeterKietzmann) [inline tests/usbus_hid/main.c:37]: In RIOT this would typically be executed from `auto_init`. I assume that you implemented it here to pass `report_desc_ctap` correct? @bergzand do we have a similar situation with other USB classes?

File: tests/usbus_hid/Makefile
Comment (Ollrogge) [reply tests/usbus_hid/Makefile]: No it is not.
Fixed it, code is now using the USB_VID and USB_PID testing variables, which are necessary.

File: tests/usbus_hid/main.c
Comment (Ollrogge) [reply tests/usbus_hid/main.c:55]: Correct. Because there are so many configuration options with HID report descriptors, I couldn't find a better solution than to simply disable auto_init_usbus and initialize it manually.

File: tests/usbus_hid/main.c
Comment (Ollrogge) [reply tests/usbus_hid/main.c:60]: I am unsure about this file in general.
It is supposed to be an interface for the usb_hid module, which handles the I/O only for this module, not for stdio like cdc_acm_stdio.c does. 

What would be the right approach for this ?

File: sys/usb/usbus/hid/hid_io.c
Comment (PeterKietzmann) [inline sys/usb/usbus/hid/hid_io.c:45]: `usb_hid_io_write`?

File: tests/usbus_hid/Makefile
Comment (PeterKietzmann) [inline tests/usbus_hid/Makefile:9]: Maybe even better `USB_VID ?= $(USB_VID_TESTING)` to allow change during `make` call.

File: tests/usbus_hid/main.c
Comment (PeterKietzmann) [reply tests/usbus_hid/main.c:60]: >  unsure about this file

Do you mean function `usb_hid_io_init` or the file *hid_io.c*?  Don't know exactly about the I/O interface, but some interface seems reasonable instead of directly having to access `isrpipe` for instance...

**Comment by PeterKietzmann** [comment]
@Ollrogge would you please fix the following:
```
	sys/include/usb/usbus/hid.h:105: warning: line is longer than 100 characters
	sys/include/usb/usbus/hid.h:106: warning: line is longer than 100 characters
	sys/include/usb/usbus/hid.h:181: warning: too many consecutive empty lines
	sys/usb/usbus/hid/hid_io.c:74: warning: no newline at end of file
	tests/usbus_hid/main.c:27: warning: line is longer than 100 characters
```
```
sys/usb/usbus/hid/hid.c:174: unhandeled  ==> unhandled
```

@bergzand would you mind to have a quick look at this PR? To me it looks fine and the test works as expected, but I haven't reviewed USB-related stuff before...

**Comment by Ollrogge** [comment]
> Hi @Ollrogge and thank you for this contribution. From what I see this implementation matches the flow used with CDC-ACM closely. The CDC-ACM module uses the tsrb and other blocking buffers to synchronize different threads writing to stdout facility.
> 
> If I understand correctly, the hid.c handles the USB HID interaction and the `hid_io.c` file can be used if a more file IO-like interaction is required. Is that right? Is it possible (and supported) to use just the hid.c code and use the callbacks there for event-driven interaction with the HID module?
> 
> Ideally I would prefer to have the tsrb code that is now in `usbus/hid/hid.c` in the `hid_io.c`here. This would also require a way to submit HID messages to the IN endpoint, I think that could be handled with a different event allocated by the sender and a callback to indicate that it is transferred to the host if required. This would remove the need for the flush and would rewrite the usbus_hid_submit to an event submission. The `usb_hid_io.c` can then wrap around this with a blocking mechanism.
> 
> One of the issues with the tsrb is that there is no way to separate different reports. I think if the calling code pushes multiple different reports to the tsrb the reports can get appended into a single USB frame possibly leading to report parsing errors on the host side. What do you think? Maybe there is a different mechanism more suitable to synchronize this.

Hi @bergzand, your understanding of the roles of ```hid.c``` and ```hid_io.c``` is correct.
I have not thought about event-driven interaction with the HID module but I like the idea.

To make sure I understand your suggestions correctly:

There would be a new event and a handler for that event in the ```hid.c``` code.
This event can be queued by calling ```usbus_event_post```.
When the event is triggered, the HID report data from the sender is written to the IN enpoint and transferred to the host.
Once the event has been processed, a callback will be called indicating to the sender, that the data has been transferred to the host.

This approach would require a new kind of ```event``` struct in order to hold a pointer to a data buffer and a function pointer to the callback right ?

Regarding the possible issue with tsrb and reports:

I am unsure if this could lead to problems. Let's say we have an endpoint size of 64 bytes and 2 different report types,
one being 32 bytes the second one being 48 bytes. Having two different report types, we can define report IDS in the HID report descriptor.
This would enable the host to know how much more bytes to expect and thefore should not result in parsing problems.

**Comment by Ollrogge** [comment]
Hi @bergzand, 
please check my latest fixup, where I implemented an event-based approach for submitting HID messages to the IN endpoint.

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:181]: Just for my understanding: if we were to implement a keyboard, this would be the part where the host would query the key events?

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:175]:
```suggestion
            else if (desc_type == USB_HID_DESCR_HID) {
```

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:128]: can we have a define for that?

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:140]: Why are there two endpoints?

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:69]: I wonder if this should be supplied by a config struct.
But that could also be done in a later PR if the HID feature gains more capabilities.

File: sys/include/usb/usbus/hid.h
Comment (benpicco) [inline sys/include/usb/usbus/hid.h:86]: should that be `const`?

**Review by benpicco — COMMENTED** [review]
sorry for the long radio silence.
I don't know much about USB or HID, but some comments:

File: sys/usb/usbus/hid/hid.c
Comment (Ollrogge) [reply sys/usb/usbus/hid/hid.c:179]: It can be, yes. The HID spec states: "Get_Report request allows the host to receive a report via the Control pipe.".

Receiving a report via the Control pipe is not the normal case though. Normally, the key events would be polled by the host
via the Interrupt pipe.

File: sys/usb/usbus/hid/hid.c
Comment (Ollrogge) [reply sys/usb/usbus/hid/hid.c:138]: This is due to the fact, that this USB HID interface was needed as base for the implementation of the CTAP2 protocol (part of the FIDO2 standard) in RIOT.
CTAP2 is mostly used for security keys, which need a OUT endpoint. 

For a keyboard, the OUT endpoint is useless. There is more code needed to analyze the HID report descriptor and 
add endpoints based on that.

File: sys/usb/usbus/hid/hid.c
Comment (Ollrogge) [reply sys/usb/usbus/hid/hid.c:69]: Yes. There is more code needed to represent and analyze HID report descriptors.

File: sys/include/usb/usbus/hid.h
Comment (Ollrogge) [reply sys/include/usb/usbus/hid.h:86]: Good idea. Made it const.

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:126]: what unit are those in?

File: sys/usb/usbus/hid/hid.c
Comment (benpicco) [inline sys/usb/usbus/hid/hid.c:135]: duplicate?

**Comment by benpicco** [comment]
We shouldn't let this got stale.
As a minimal implementation, this works and can be updated for more use-cases with follow-up PRs.

Please squash!

File: Makefile.dep
Comment (dylad) [inline Makefile.dep:974]: You shoud move this to sys/Makefile.dep after rebasing this PR on top of master.

File: sys/usb/usbus/hid/hid.c
Comment (dylad) [inline sys/usb/usbus/hid/hid.c:81]: I think it will be better and less error prone to have something like
`hid->ep_name `

File: tests/usbus_hid/main.c
Comment (dylad) [inline tests/usbus_hid/main.c:10]: this field should be update.

File: sys/usb/usbus/hid/hid_io.c
Comment (dylad) [inline sys/usb/usbus/hid/hid_io.c:48]: same comment here.
It would be better to avoid ep->next->ep but rather directly point to the right endpoint.
You could add a pointer to the rightful endpoint in your usbus_hid_device_t struct.

**Review by dylad — CHANGES_REQUESTED** [review]
Sorry for jumping in late !
Overall looks good to me, only a few changes required.
Thanks for this amazing PR !

File: sys/usb/usbus/hid/hid_io.c
Comment (Ollrogge) [reply sys/usb/usbus/hid/hid_io.c:48]: I don't know if this is a good idea, because the HID endpoint can have a variable amount of endpoints.

File: tests/usbus_hid/main.c
Comment (Ollrogge) [reply tests/usbus_hid/main.c]: change it to tests ? I that what you meant?

File: Makefile.dep
Comment (Ollrogge) [reply Makefile.dep]: done

File: tests/usbus_hid/main.c
Comment (dylad) [reply tests/usbus_hid/main.c]: Yes

File: sys/usb/usbus/hid/hid_io.c
Comment (dylad) [reply sys/usb/usbus/hid/hid_io.c:48]: I understand what you mean but on the other side, if we add or modify the endpoints registration during the hid initialization, we might break this.

File: sys/usb/usbus/hid/hid.c
Comment (dylad) [inline sys/usb/usbus/hid/hid.c:120]:
```suggestion
    hid->ep_in = usbus_add_endpoint(usbus, &hid->iface,
```

File: sys/usb/usbus/hid/hid.c
Comment (dylad) [inline sys/usb/usbus/hid/hid.c:132]:
```suggestion
     hid->ep_out = usbus_add_endpoint(usbus, &hid->iface,
```

**Comment by Ollrogge** [comment]
Hey @dylad, is there anything missing ?

**Comment by dylad** [comment]
Looks like my comments were addressed.
@bergzand @benpicco @PeterKietzmann How about yours ?

**Comment by benpicco** [comment]
looks good to me, one more thing though: Please change the commit message to something like `usbus/hid: add minimal implementation`

**Comment by Ollrogge** [comment]
@benpicco changed the commit message.

**Comment by bergzand** [comment]
>  @bergzand @benpicco @PeterKietzmann How about yours ?

My comments have been addressed!

**Comment by dylad** [comment]
@Ollrogge Could you rebase one last time please ?
This will fix the required check-pr / check-commits

**Comment by Ollrogge** [comment]
@dylad Done :)

**Review by dylad — APPROVED** [review]
ACK.
This PR is in good shape and works as advertised. USB looks good to me.
We can add keyboard or fancy stuff support in followup PRs.
@bergzand do you mind dismiss your changes request please ?

**Comment by PeterKietzmann** [comment]
> @bergzand @benpicco @PeterKietzmann How about yours ?

Looks good to me and I see many green lights. Only waiting for @bergzand to approve.

**Review by bergzand — APPROVED** [review]
ACK, sorry for blocking!

**Comment by dylad** [comment]
Here we go !

File: sys/include/usb/usbus/hid.h
Comment (aabadie) [inline sys/include/usb/usbus/hid.h:104]: This function and `usbus_hid_flush` below are not implemented. @Ollrogge do you plan to add them ?

Comment (Ollrogge) [reply sys/include/usb/usbus/hid.h:104]: This must be a left over from my initial implementation. This function is not needed anymore. I will remove the declaration.

**Pull request merged**
